draft-ietf-tls-tls13.md

```diff
@@ -4696,6 +4641,193 @@ a context in all exporter computations, though the value could be empty.
+connections when compared to self-encrypted tickets. Unlike
+session databases, session tickets can successfully do PSK-based
+session establishment even without consistent storage, though when
+0-RTT is allowed they still require consistent storage for anti-replay
+of 0-RTT data, as detailed in the following
+section.
+
+## Client Hello Recording
+
+An alternative form of anti-replay is to record a unique value derived
+from the ClientHello (generally either the random value or the PSK
+binder) and reject duplicates. Recording all ClientHellos causes state
+to grow without bound, but a server can instead record ClientHellos within
+a given time window and use the "obfuscated_ticket_age" to ensure that
+tickets aren't reused outside that window.
+
+In order to implement this, when a ClientHello is received, the server
+first verifies the PSK binder as described
+{{pre-shared-key-extension}}. It then computes the
+expected_arrival_time as described in the next section and and rejects
+0-RTT if it is outside the recording window, falling back to the
+1-RTT handshake.
+
+If the expected arrival time is in the window, then the server
+checks to see if it has recorded a matching ClientHello. It
+either aborts the handshake with an "illegal_parameter" alert
+or accepts the PSK but reject 0-RTT. If no matching ClientHello
+is found, then it stores the ClientHello as long as the
+expected_arrival_time is inside the window and accepts 0-RTT.
+Servers MAY also implement data stores with false positives, such as
+Bloom filters, in which case they MUST respond to apparent replay by
+rejecting 0-RTT but MUST NOT abort the handshake.
+
+The server MUST derive the storage key only from validated sections
+of the ClientHello. If the ClientHello contains multiple
+PSK identities, then an attacker can create multiple ClientHellos
+with different binder values for the less-preferred identity on the
+assumption that the server will not verify it. I.e., if the
+client sends PSKs A and B but the server prefers A, then the
+attacker can change the binder for B without affecting the binder
+for A. If the validated binder or the ClientHello.random
+are used for this purpose, then this attack is not possible.
+
+Because this mechanism does not require storing all outstanding
+tickets, it may be easier to implement in distributed systems with
+high rates of resumption and 0-RTT, at the cost of potentially
+weaker anti-replay defense because of the difficulty of reliably
+storing and retrieving the received ClientHello messages.
+In many such systems, it is impractical to have globally
+consistent storage of all the received ClientHellos.
+In this case, the stronger design is to have a single storage zone be
+authoritative for a given ticket and refuse 0-RTT for that
+ticket in any other zone. This approach prevents simple
+replay by the attacker because only one zone will accept
+0-RTT data. A weaker design is to implement separate storage for
```

Comment:
Oh, describing multiple options is totally within editor discretion.
The main thing I wanted to comment about in this space was that starting a sentence "In this case, the stronger design" is a strange construction (in that it uses the definite article before it's clear what's being referred to), which it looks like you changed already.  (It's fine to say "a weaker design" without a preceeding "stronger design".)
So, I am belatedly realizing that you are only responding to my "mixed feelings" remark, which was not intended to be anywhere close to blocking, so I'll just shut up now; sorry for the noise.